src/mlpack/core/cv/k_fold_cv.hpp

```diff
@@ -0,0 +1,219 @@
+   */
+  template<typename DataArgsTupleT,
+           typename = typename std::enable_if<
+               std::tuple_size<DataArgsTupleT>::value == 3>::type,
+           typename = void>
+  void Init(const DataArgsTupleT& dataArgsTuple);
+
+  /**
+   * Initialize the given destination matrix with the given source joined with
+   * its first k - 2 bins.
+   */
+  template<typename SourceType, typename DestinationType>
+  void InitKFoldCVMat(const SourceType& source, DestinationType& destination);
+
+  /**
+   * Train and run evaluation in the case of non-weighted learning.
+   */
+  template<typename...MLAlgorithmArgs,
+           bool Enabled = !Base::MIE::SupportsWeights,
+           typename = typename std::enable_if<Enabled>::type>
+  double TrainAndEvaluate(const MLAlgorithmArgs& ...mlAlgorithmArgs);
+
+  /**
+   * Train and run evaluation in the case of supporting weighted learning.
+   */
+  template<typename...MLAlgorithmArgs,
+           bool Enabled = Base::MIE::SupportsWeights,
+           typename = typename std::enable_if<Enabled>::type,
+           typename = void>
+  double TrainAndEvaluate(const MLAlgorithmArgs& ...mlAlgorithmArgs);
+
+  /**
+   * Calculate the index of the first column of the ith validation subset.
+   *
+   * We take the ith validation subset after the ith training subset if
+   * i < k - 1 and before it otherwise.
+   */
+  size_t ValidationSubsetFirtsCol(const size_t i)
+  {
+    return i < k - 1? binSize * i + trainingSubsetSize : binSize * (i - 1);
+  }
+
+  /**
+   * Get the ith training subset from a variable of a matrix type.
+   */
+  template<typename ElementType>
+  arma::Mat<ElementType> GetTrainingSubset(arma::Mat<ElementType>& m,
+                                           const size_t i)
+  {
+    return arma::Mat<ElementType>(m.colptr(binSize * i), m.n_rows,
+        trainingSubsetSize, false, true);
+  }
+
+  /**
+   * Get the ith training subset from a variable of a row type.
+   */
+  template<typename ElementType>
+  arma::Row<ElementType> GetTrainingSubset(arma::Row<ElementType>& r,
+                                           const size_t i)
+  {
+    return arma::Row<ElementType>(r.colptr(binSize * i), trainingSubsetSize,
+        false, true);
+  }
+
+  /**
+   * Get the ith validation subset from a variable of a matrix type.
+   */
+  template<typename ElementType>
+  arma::Mat<ElementType> GetValidationSubset(arma::Mat<ElementType>& m,
+                                             const size_t i)
+  {
+    return arma::Mat<ElementType>(m.colptr(ValidationSubsetFirtsCol(i)),
+        m.n_rows, binSize, false, true);
+  }
+
+  /**
+   * Get the ith validation subset from a variable of a row type.
+   */
+  template<typename ElementType>
+  arma::Row<ElementType> GetValidationSubset(arma::Row<ElementType>& r,
+                                             const size_t i)
+  {
+    return arma::Row<ElementType>(r.colptr(ValidationSubsetFirtsCol(i)),
+        binSize, false, true);
+  }
```

Comment:
Since there's already a k_fold_cv_impl.hpp, it may be better just to split the
implementation of these simple functions into that file (since they don't fit
all on one line), and then just mark them 'inline' in this file.  That way
someone who reads the header file gets a quick idea of all the methods in the
class, but doesn't need to think about implementation at all.